### PR TITLE
Display splits in previews

### PR DESCRIPTION
**Pull request by jfdoming**
Following up on https://github.com/actualbudget/actual/issues/2379#issuecomment-2136242633.

There are some design decisions I'd like feedback on. In particular, Actual historically used the "category" column for the schedule status. That gets strange with splits because we also use that column for splits. I tentatively used the below UI:
image
image

We could also do something like this, although it looks a bit odd when the split is not hovered:
image
image

Open to suggestions or better alternatives.

### CODE REVIEW

**Comment by jfdoming** [comment]
note: updated vrt because I slightly changed padding on the mobile view to be more consistent across preview/non-preview transactions.

**Comment by youngcw** [comment]
The upcoming/due/missed labels seem to have gotten lost on most schedules. I do wonder if it would be better to have split schedules always collapsed unless manually expanded?  I like that it can be opened, but I think that having all the children lines present all the time could quickly make for a lot of lines on the page.

**Comment by psybers** [comment]
I agree, having upcoming with splits collapsed by default seems reasonable.  Generally, it is more important to know the transaction is coming, not how that transaction will be split up.

I think the bug is if an upcoming has no split, it no longer says upcoming/missed.

**Comment by jfdoming** [comment]
> The upcoming/due/missed labels seem to have gotten lost on most schedules. I do wonder if it would be better to have split schedules always collapsed unless manually expanded? I like that it can be opened, but I think that having all the children lines present all the time could quickly make for a lot of lines on the page.

Hmm, so my thinking is that having splits collapsed by default could make it hard to tell what the split is. For example, here is a screenshot of a scheduled split expanded and collapsed:
image
image
With the split collapsed you have almost no information about what the transaction actually is, since now we display "Split" for the payee instead of the actual payee. I suppose you could add a note to display to clarify? But I'm not sure it's the best UX. Maybe the correct solution is to revert to showing the parent payee instead?

**Comment by psybers** [comment]
> Maybe the correct solution is to revert to showing the parent payee instead?

@jfdoming I think in this instance, it makes sense to show the payee instead of 'Split' and then collapse it. It already says split on the category column so you know it is one.

One other thing: should the collapse state follow the 'expand split transactions' button?  In other words, should it collapse only if the user has *all* splits collapse?

**Comment by joel-jeremy** [comment]
I agree, having the scheduled split collapsed by default seems reasonable. Users can always choose to expand if they need more info, what's more important is that the transaction gets in their radars. Just an idea - what if we override the Split payee button on preview transactions to also collapse/expand like the category arrow?

**Comment by jfdoming** [comment]
> > Maybe the correct solution is to revert to showing the parent payee instead?
> 
> @jfdoming I think in this instance, it makes sense to show the payee instead of 'Split' and then collapse it. It already says split on the category column so you know it is one.
> 
> One other thing: should the collapse state follow the 'expand split transactions' button? In other words, should it collapse only if the user has _all_ splits collapse?

👋 for your first comment, which payee should get used? My understanding is that we were moving away from showing payees on the parent transaction (starting in https://github.com/actualbudget/actual/pull/2805), and if the child transactions have multiple payees, I'm not sure which one would be best to show

For the second comment, I believe this is already the behaviour. I think @youngcw and @joel-jeremy were suggesting that we _ignore_ that global toggle and instead always collapse these splits by default. Open to discussion!

**Comment by jfdoming** [comment]
> I agree, having the scheduled split collapsed by default seems reasonable. Users can always choose to expand if they need more info, what's more important is that the transaction gets in their radars. Just an idea - what if we override the Split payee button on preview transactions to also collapse/expand like the category arrow?

@joel-jeremy nice idea! Is there a reason that is not already the behaviour on non-preview transactions? Also wondering if you think it's worth having _both_ the clickable Split payee and arrow, or just the Split payee.

**Comment by joel-jeremy** [comment]
@jfdoming The split payee button is there for non preview transactions to easily set payee for all its child transactions.

There is an idea that came up on the split payee on mobile PR where children payees are shown as `My payee (+3)` or only `My payee` if all children payees are the same. Maybe we can do that here too when the split transaction is collapsed? WDYT?

**Comment by jfdoming** [comment]
> @jfdoming The split payee button is there for non preview transactions to easily set payee for all its child transactions.
> 
> There is an idea that came up on the split payee on mobile PR where children payees are shown as `My payee (+3)` or only `My payee` if all children payees are the same. Maybe we can do that here too when the split transaction is collapsed? WDYT?

@joel-jeremy it's a nice idea! I moved that change to https://github.com/actualbudget/actual/pull/3049 since it seems a bit unrelated here. Re: auto-collapsing or not collapsing, I think it boils down to:
- Splits in previews follow the global "auto-expand" setting; i.e., this button: 
image

- OR, splits in previews are default collapsed, ignoring the "auto-expand" setting

What do you think is best here? IMO following the global setting is more intuitive, but defer to you. Happy to make either change

**Comment by psybers** [comment]
I honestly could see either argument.  Me personally, I expand splits but would prefer them not expanded in an upcoming transaction.  I don't really care about how upcoming transactions are split up, just that I have $X coming out of Y account in a few days.

**Comment by joel-jeremy** [comment]
Maybe we could collapse the upcoming splits by default (ignoring the global settings) on initial load of the page. But once user toggle the setting, it should follow that.

**Comment by jfdoming** [comment]
> Maybe we could collapse the upcoming splits by default (ignoring the global settings) on initial load of the page. But once user toggle the setting, it should follow that.

That makes sense to me! I'll try to figure that out later today after work

**Comment by jfdoming** [comment]
@joel-jeremy @psybers should be good to go now! let me know if this behaves how you were thinking

File: packages/desktop-client/src/hooks/usePreviewTransactions.ts
Comment (joel-jeremy) [inline packages/desktop-client/src/hooks/usePreviewTransactions.ts:14]: Is it possible to call this outside the hook? Calling it here seems out of place since the hook is just intended to retrieve the preview transactions.

File: packages/desktop-client/src/hooks/usePreviewTransactions.ts
Comment (joel-jeremy) [inline packages/desktop-client/src/hooks/usePreviewTransactions.ts:17]: Can we use `usePrevious` here?

**Review by joel-jeremy — COMMENTED** [review]
Overall LGTM! Thank you for all your effort on this PR! Just a few comments on the `usePreviewTransactions` hook.

**Comment by psybers** [comment]
@jfdoming when I click the icons in the red box:

image

it lets me edit the schedule.  For the split, I can no longer do that?

**Comment by psybers** [comment]
Also the payee now is just 'Split'? So we can no longer view the payee?

**Comment by psybers** [comment]
image

The "Split" text renders like a button (note the hover) but clicking does nothing.

**Comment by jfdoming** [comment]
@psybers 
> For the split, I can no longer do that?
Nice catch, let me think on how best to handle this

> Also the payee now is just 'Split'? So we can no longer view the payee?
That's a change from a previous PR (current behaviour for split transactions on `master`). It should be fixed by https://github.com/actualbudget/actual/pull/3049

> The "Split" text renders like a button (note the hover) but clicking does nothing.
Fair point, for preview transactions we should probably make it not clickable

File: packages/desktop-client/src/hooks/usePreviewTransactions.ts
Comment (jfdoming) [reply packages/desktop-client/src/hooks/usePreviewTransactions.ts:14]: Yeah, it's a bit out of place for sure. I went with this approach because the logic needs to run when the preview transactions are _loaded_ rather than in a render somewhere (I tried an approach with a `useEffect` before this and it got extremely messy). One mitigation maybe is to rename from `collapseTransactions` to `onLoad`. What do you think?

File: packages/desktop-client/src/hooks/usePreviewTransactions.ts
Comment (jfdoming) [reply packages/desktop-client/src/hooks/usePreviewTransactions.ts:13]: The data in this hook is now lazily loaded (since it has to make calls to async APIs), but the caller needs the data (or empty) non-lazily. I went with this approach so we can return the data via a `setState`. Does that make sense?

File: packages/desktop-client/src/hooks/usePreviewTransactions.ts
Comment (jfdoming) [reply packages/desktop-client/src/hooks/usePreviewTransactions.ts:17]: I think `usePrevious` always returns the value on the previous render, whereas here we need the value from the last time `scheduleData` was changed. Maybe a name like `currentScheduleData` would be more accurate to the intent. What do you think of that name?

**Comment by youngcw** [comment]
@jfdoming Can you fix the conflicts on this?

**Review by matt-fidd — APPROVED** [review]
Looks good!

**Comment by jfdoming** [comment]
Any more thoughts @psybers @youngcw? It would be nice to get this merged since it's been around for a while 😅 but wanted to give a chance to comment once more before I hit send

**Comment by psybers** [comment]
@jfdoming None from me, it looks good! I did not review the code, just functionality.

**Comment by youngcw** [comment]
It seems fine to me

**Pull request merged**
